leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/DefaultLwM2mNodeDecoder.java

```diff
@@ -0,0 +1,158 @@
+    public <T extends LwM2mNode> T decode(byte[] content, ContentFormat format, LwM2mPath path, LwM2mModel model,
+            Class<T> nodeClass) throws InvalidValueException {
+
+        LOG.debug("Decoding value for path {} and format {}: {}", path, format, content);
+        Validate.notNull(path);
+
+        // If no format is given, guess the best one to use.
+        if (format == null) {
+            format = guessContentType(path, model);
+        }
+
+        // Decode content.
+        switch (format.getCode()) {
+        case ContentFormat.TEXT_CODE:
+            return (T) LwM2mNodeTextDecoder.decode(content, path, model);
+        case ContentFormat.TLV_CODE:
+            return LwM2mNodeTlvDecoder.decode(content, path, model, nodeClass);
+        case ContentFormat.OPAQUE_CODE:
+            return (T) LwM2mNodeOpaqueDecoder.decode(content, path, model);
+        case ContentFormat.JSON_CODE:
+            return LwM2mNodeJsonDecoder.decode(content, path, model, nodeClass);
+        case ContentFormat.LINK_CODE:
+            throw new UnsupportedOperationException("Content format " + format + " not yet implemented '" + path + "'");
+        }
+        return null;
+    }
+
+    @Override
+    public List<TimestampedLwM2mNode> decodeTimestampedData(byte[] content, ContentFormat format, LwM2mPath path,
+            LwM2mModel model) throws InvalidValueException {
+        LOG.debug("Decoding value for path {} and format {}: {}", path, format, content);
+        Validate.notNull(path);
+
+        // If no format is given, guess the best one to use.
+        if (format == null) {
+            format = guessContentType(path, model);
+        }
+
+        // Decode content.
+        switch (format.getCode()) {
+        case ContentFormat.TEXT_CODE:
+            return toTimestampedNodes(LwM2mNodeTextDecoder.decode(content, path, model));
+        case ContentFormat.TLV_CODE:
+            return toTimestampedNodes(LwM2mNodeTlvDecoder.decode(content, path, model, nodeClassFromPath(path)));
+        case ContentFormat.OPAQUE_CODE:
+            return toTimestampedNodes(LwM2mNodeOpaqueDecoder.decode(content, path, model));
+        case ContentFormat.JSON_CODE:
+            return LwM2mNodeJsonDecoder.decodeTimestampedData(content, path, model, nodeClassFromPath(path));
+        case ContentFormat.LINK_CODE:
+            throw new UnsupportedOperationException("Content format " + format + " not yet implemented '" + path + "'");
+        }
+        return null;
+    }
+
+    private static List<TimestampedLwM2mNode> toTimestampedNodes(LwM2mNode node) {
+        if (node == null)
+            return Collections.emptyList();
+
+        ArrayList<TimestampedLwM2mNode> timestampedNodes = new ArrayList<>(1);
+        timestampedNodes.add(new TimestampedLwM2mNode(null, node));
+        return Collections.unmodifiableList(timestampedNodes);
+    }
+
+    private static Class<? extends LwM2mNode> nodeClassFromPath(LwM2mPath path) {
```

Comment:
It would make sense to make this method public since it will probably be needed to build your own decoder.